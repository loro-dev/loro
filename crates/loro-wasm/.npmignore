/target
*.log
*.ipynb
flamegraph.svg
target
dhat-heap.json
.DS_Store
node_modules/
deno_test/
src/
wasm-size/
web-test/
pkg/
web/
