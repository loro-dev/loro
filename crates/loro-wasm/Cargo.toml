[package]
name = "loro-wasm"
version = "0.1.0"
edition = "2021"

[lib]
crate-type = ["cdylib", "rlib"]

[dependencies]
js-sys = "0.3.60"
loro-internal = { path = "../loro-internal", features = ["wasm"] }
wasm-bindgen = "0.2.83"
serde-wasm-bindgen = { version = "0.4.5" }
console_error_panic_hook = { version = "0.1.6", optional = true }
tracing-wasm = "0.2.1"
tracing = "0.1.37"

[features]
default = ["console_error_panic_hook"]
