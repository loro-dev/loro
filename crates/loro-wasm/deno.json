{
  "tasks": {
    "release": "deno run -A ./scripts/build.ts release",
    "dev": "deno run -A ./scripts/build.ts dev"
  },
  "test": {
    "files": {
      "include": ["deno_test/"]
    }
  }
}
