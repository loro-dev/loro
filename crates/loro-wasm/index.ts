export * from "loro-wasm";
export type * from "loro-wasm";
import {
    AwarenessWasm,
    PeerID,
    Container,
    ContainerID,
    ContainerType,
    LoroCounter,
    LoroDoc,
    LoroList,
    LoroMap,
    LoroText,
    LoroTree,
    OpId,
    Value,
    AwarenessListener,
} from "loro-wasm";

/**
 * @deprecated Please use LoroDoc
 */
export class Loro extends LoroDoc { }

const CONTAINER_TYPES = [
    "Map",
    "Text",
    "List",
    "Tree",
    "MovableList",
    "Counter",
];

export function isContainerId(s: string): s is ContainerID {
    return s.startsWith("cid:");
}

/**  Whether the value is a container.
 *
 * # Example
 *
 * ```ts
 * const doc = new LoroDoc();
 * const map = doc.getMap("map");
 * const list = doc.getList("list");
 * const text = doc.getText("text");
 * isContainer(map); // true
 * isContainer(list); // true
 * isContainer(text); // true
 * isContainer(123); // false
 * isContainer("123"); // false
 * isContainer({}); // false
 * ```
 */
export function isContainer(value: any): value is Container {
    if (typeof value !== "object" || value == null) {
        return false;
    }

    const p = Object.getPrototypeOf(value);
    if (p == null || typeof p !== "object" || typeof p["kind"] !== "function") {
        return false;
    }

    return CONTAINER_TYPES.includes(value.kind());
}


/**  Get the type of a value that may be a container.
 *
 * # Example
 *
 * ```ts
 * const doc = new LoroDoc();
 * const map = doc.getMap("map");
 * const list = doc.getList("list");
 * const text = doc.getText("text");
 * getType(map); // "Map"
 * getType(list); // "List"
 * getType(text); // "Text"
 * getType(123); // "Json"
 * getType("123"); // "Json"
 * getType({}); // "Json"
 * ```
 */
export function getType<T>(
    value: T,
): T extends LoroText ? "Text"
    : T extends LoroMap<any> ? "Map"
    : T extends LoroTree<any> ? "Tree"
    : T extends LoroList<any> ? "List"
    : T extends LoroCounter ? "Counter"
    : "Json" {
    if (isContainer(value)) {
        return value.kind() as unknown as any;
    }

    return "Json" as any;
}


export function newContainerID(id: OpId, type: ContainerType): ContainerID {
    return `cid:${id.counter}@${id.peer}:${type}`;
}

export function newRootContainerID(
    name: string,
    type: ContainerType,
): ContainerID {
    return `cid:root-${name}:${type}`;
}



/**
 * Awareness is a structure that allows to track the ephemeral state of the peers.
 *
 * If we don't receive a state update from a peer within the timeout, we will remove their state.
 * The timeout is in milliseconds. This can be used to handle the off-line state of a peer.
 */
export class Awareness<T extends Value = Value> {
    inner: AwarenessWasm<T>;
    private peer: PeerID;
    private timer: number | undefined;
    private timeout: number;
    private listeners: Set<AwarenessListener> = new Set();
    constructor(peer: PeerID, timeout: number = 30000) {
        this.inner = new AwarenessWasm(peer, timeout);
        this.peer = peer;
        this.timeout = timeout;
    }

    apply(bytes: Uint8Array, origin = "remote") {
        const { updated, added } = this.inner.apply(bytes);
        this.listeners.forEach((listener) => {
            listener({ updated, added, removed: [] }, origin);
        });

        this.startTimerIfNotEmpty();
    }

    setLocalState(state: T) {
        const wasEmpty = this.inner.getState(this.peer) == null;
        this.inner.setLocalState(state);
        if (wasEmpty) {
            this.listeners.forEach((listener) => {
                listener(
                    { updated: [], added: [this.inner.peer()], removed: [] },
                    "local",
                );
            });
        } else {
            this.listeners.forEach((listener) => {
                listener(
                    { updated: [this.inner.peer()], added: [], removed: [] },
                    "local",
                );
            });
        }

        this.startTimerIfNotEmpty();
    }

    getLocalState(): T | undefined {
        return this.inner.getState(this.peer);
    }

    getAllStates(): Record<PeerID, T> {
        return this.inner.getAllStates();
    }

    encode(peers: PeerID[]): Uint8Array {
        return this.inner.encode(peers);
    }

    encodeAll(): Uint8Array {
        return this.inner.encodeAll();
    }

    addListener(listener: AwarenessListener) {
        this.listeners.add(listener);
    }

    removeListener(listener: AwarenessListener) {
        this.listeners.delete(listener);
    }

    peers(): PeerID[] {
        return this.inner.peers();
    }

    destroy() {
        clearInterval(this.timer);
        this.listeners.clear();
    }

    private startTimerIfNotEmpty() {
        if (this.inner.isEmpty() || this.timer != null) {
            return;
        }

        this.timer = setInterval(() => {
            const removed = this.inner.removeOutdated();
            if (removed.length > 0) {
                this.listeners.forEach((listener) => {
                    listener({ updated: [], added: [], removed }, "timeout");
                });
            }
            if (this.inner.isEmpty()) {
                clearInterval(this.timer);
                this.timer = undefined;
            }
        }, this.timeout / 2) as unknown as number;
    }
}

LoroDoc.prototype.toJsonWithReplacer = function (replacer: (key: string | number, value: Value | Container) => Value | Container | undefined) {
    const doc = this;
    const m = (key: string | number, value: Value): Value | undefined => {
        if (typeof value === "string") {
            if (isContainerId(value)) {
                const container = doc.getContainerById(value);
                if (container == null) {
                    throw new Error(`ContainerID not found: ${value}`);
                }

                const ans = replacer(key, container);
                if (ans === container) {
                    const ans = container.getShallowValue();
                    if (typeof ans === "object") {
                        return run(ans as any);
                    }

                    return ans;
                }

                if (isContainer(ans)) {
                    throw new Error("Using new container is not allowed in toJsonWithReplacer");
                }

                return ans;
            }
        }

        const ans = replacer(key, value);
        if (isContainer(ans)) {
            throw new Error("Using new container is not allowed in toJsonWithReplacer");
        }

        return ans;
    }

    const run = (layer: Record<string, Value> | Value[]): Value => {
        if (Array.isArray(layer)) {
            return layer.map((item, index) => {
                return m(index, item);
            }).filter((item): item is NonNullable<typeof item> => item !== undefined);
        }

        const result: Record<string, Value> = {};
        for (const [key, value] of Object.entries(layer)) {
            const ans = m(key, value);
            if (ans !== undefined) {
                result[key] = ans;
            }
        }

        return result;
    }

    const layer = this.getShallowValue();
    return run(layer);
}

