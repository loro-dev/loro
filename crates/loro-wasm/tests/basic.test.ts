import { describe, expect, expectTypeOf, it } from "vitest";
import {
  Container,
  Diff,
  getType,
  isContainer,
  LoroDoc,
  LoroList,
  LoroMap,
  LoroText,
  LoroTree,
  VersionVector,
  MapDiff,
  TextDiff,
  Frontiers,
  encodeFrontiers,
  decodeFrontiers,
} from "../bundler/index";
import { ContainerID } from "loro-wasm";

it("basic example", () => {
  const doc = new LoroDoc();
  const list = doc.getList("list");
  list.insert(0, "A");
  list.insert(1, "B");
  list.insert(2, "C");

  const map = doc.getMap("map");
  // map can only has string key
  map.set("key", "value");
  expect(doc.toJSON()).toStrictEqual({
    list: ["A", "B", "C"],
    map: { key: "value" },
  });

  // delete 2 elements at index 0
  list.delete(0, 2);
  expect(doc.toJSON()).toStrictEqual({
    list: ["C"],
    map: { key: "value" },
  });

  // Insert a text container to the list
  const text = list.insertContainer(0, new LoroText());
  text.insert(0, "Hello");
  text.insert(0, "Hi! ");

  // delete 1 element at index 0
  expect(doc.toJSON()).toStrictEqual({
    list: ["Hi! Hello", "C"],
    map: { key: "value" },
  });

  // Insert a list container to the map
  const list2 = map.setContainer("test", new LoroList());
  list2.insert(0, 1);
  expect(doc.toJSON()).toStrictEqual({
    list: ["Hi! Hello", "C"],
    map: { key: "value", test: [1] },
  });
});

it("get or create on Map", () => {
  const docA = new LoroDoc();
  const map = docA.getMap("map");
  const container = map.getOrCreateContainer("list", new LoroList());
  container.insert(0, 1);
  container.insert(0, 2);
  const text = map.getOrCreateContainer("text", new LoroText());
  text.insert(0, "Hello");
  expect(docA.toJSON()).toStrictEqual({
    map: { list: [2, 1], text: "Hello" },
  });
});

it("basic sync example", () => {
  const docA = new LoroDoc();
  const docB = new LoroDoc();
  const listA = docA.getList("list");
  listA.insert(0, "A");
  listA.insert(1, "B");
  listA.insert(2, "C");
  // B import the ops from A
  docB.import(docA.exportFrom());
  expect(docB.toJSON()).toStrictEqual({
    list: ["A", "B", "C"],
  });

  const listB = docB.getList("list");
  // delete 1 element at index 1
  listB.delete(1, 1);
  // A import the ops from B
  docA.import(docB.exportFrom(docA.version()));
  // list at A is now ["A", "C"], with the same state as B
  expect(docA.toJSON()).toStrictEqual({
    list: ["A", "C"],
  });
  expect(docA.toJSON()).toStrictEqual(docB.toJSON());
});

it("basic events", () => {
  const doc = new LoroDoc();
  doc.subscribe((event) => { });
  const list = doc.getList("list");
});

describe("list", () => {
  it("insert containers", () => {
    const doc = new LoroDoc();
    const list = doc.getList("list");
    const map = list.insertContainer(0, new LoroMap());
    map.set("key", "value");
    const v = list.get(0) as LoroMap;
    expect(v instanceof LoroMap).toBeTruthy();
    expect(v.toJSON()).toStrictEqual({ key: "value" });
  });

  it("toArray", () => {
    const doc = new LoroDoc();
    const list = doc.getList("list");
    list.insert(0, 1);
    list.insert(1, 2);
    expect(list.toArray()).toStrictEqual([1, 2]);
    list.insertContainer(2, new LoroText());
    const t = list.toArray()[2];
    expect(isContainer(t)).toBeTruthy();
    expect(getType(t)).toBe("Text");
    expect(getType(123)).toBe("Json");
  });
});

describe("map", () => {
  it("get child container", () => {
    const doc = new LoroDoc();
    const map = doc.getMap("map");
    const list = map.setContainer("key", new LoroList());
    list.insert(0, 1);
    expect(map.get("key") instanceof LoroList).toBeTruthy();
    expect((map.get("key") as LoroList).toJSON()).toStrictEqual([1]);
  });

  it("set large int", () => {
    const doc = new LoroDoc();
    const map = doc.getMap("map");
    map.set("key", 2147483699);
    expect(map.get("key")).toBe(2147483699);
  });
});

describe("import", () => {
  it("pending and import status", () => {
    const a = new LoroDoc();
    a.setPeerId(0);
    a.getText("text").insert(0, "a");
    const b = new LoroDoc();
    b.setPeerId(1);
    b.import(a.exportFrom());
    b.getText("text").insert(1, "b");
    const c = new LoroDoc();
    c.setPeerId(2);
    c.import(b.exportFrom());
    c.getText("text").insert(2, "c");

    // c export from b's version, which cannot be imported directly to a.
    // This operation is pending.
    const status = a.import(c.exportFrom(b.version()));
    const pending = new Map();
    pending.set("2", { start: 0, end: 1 });
    expect(status).toStrictEqual({ success: new Map(), pending });
    expect(a.getText("text").toString()).toBe("a");

    // a import the missing ops from b. It makes the pending operation from c valid.
    const status2 = a.import(b.exportFrom(a.version()));
    pending.set("1", { start: 0, end: 1 });
    expect(status2).toStrictEqual({ success: pending, pending: null });
    expect(a.getText("text").toString()).toBe("abc");
  });

  it("import by frontiers", () => {
    const a = new LoroDoc();
    a.getText("text").insert(0, "a");
    const b = new LoroDoc();
    b.import(a.exportFrom());
    b.getText("text").insert(1, "b");
    b.getList("list").insert(0, [1, 2]);
    const updates = b.exportFrom(b.frontiersToVV(a.frontiers()));
    a.import(updates);
    expect(a.toJSON()).toStrictEqual(b.toJSON());
  });

  it("from snapshot", () => {
    const a = new LoroDoc();
    a.getText("text").insert(0, "hello");
    const bytes = a.exportSnapshot();
    const b = LoroDoc.fromSnapshot(bytes);
    b.getText("text").insert(0, "123");
    expect(b.toJSON()).toStrictEqual({ text: "123hello" });
  });

  it("importBatch Error #181", () => {
    const docA = new LoroDoc();
    const updateA = docA.exportSnapshot();
    const docB = new LoroDoc();
    docB.importUpdateBatch([updateA]);
    docB.getText("text").insert(0, "hello");
    docB.commit();
  });
});

describe("map", () => {
  it("keys", () => {
    const doc = new LoroDoc();
    const map = doc.getMap("map");
    map.set("foo", "bar");
    map.set("baz", "bar");
    const entries = map.keys();
    expect(entries).toStrictEqual(["baz", "foo"]);
  });

  it("values", () => {
    const doc = new LoroDoc();
    const map = doc.getMap("map");
    map.set("foo", "bar");
    map.set("baz", "bar");
    const entries = map.values();
    expect(entries).toStrictEqual(["bar", "bar"]);
  });

  it("entries", () => {
    const doc = new LoroDoc();
    const map = doc.getMap("map");
    map.set("foo", "bar");
    map.set("baz", "bar");
    map.set("new", 11);
    map.delete("new");
    const entries = map.entries();
    expect(entries).toStrictEqual([
      ["baz", "bar"],
      ["foo", "bar"],
    ]);
  });

  it("entries should return container handlers", () => {
    const doc = new LoroDoc();
    const map = doc.getMap("map");
    map.setContainer("text", new LoroText());
    map.set("foo", "bar");
    const entries = map.entries();
    expect((entries[1][1]! as Container).kind() === "Text").toBeTruthy();
  });
});

it("handlers should still be usable after doc is dropped", () => {
  const doc = new LoroDoc();
  const text = doc.getText("text");
  const list = doc.getList("list");
  const map = doc.getMap("map");
  doc.free();
  text.insert(0, "123");
  expect(text.toString()).toBe("123");
  list.insert(0, 1);
  expect(list.toJSON()).toStrictEqual([1]);
  map.set("k", 8);
  expect(map.toJSON()).toStrictEqual({ k: 8 });
});

it("get change with given lamport", () => {
  const doc1 = new LoroDoc();
  doc1.setPeerId(1);
  const doc2 = new LoroDoc();
  doc2.setPeerId(2);
  doc1.getText("text").insert(0, "01234");
  doc2.import(doc1.exportFrom());
  doc2.getText("text").insert(0, "56789");
  doc1.import(doc2.exportFrom());
  doc1.getText("text").insert(0, "01234");
  doc1.commit();
  {
    const change = doc1.getChangeAtLamport("1", 1)!;
    expect(change.lamport).toBe(0);
    expect(change.peer).toBe("1");
    expect(change.length).toBe(5);
  }
  {
    const change = doc1.getChangeAtLamport("1", 7)!;
    expect(change.lamport).toBe(0);
    expect(change.peer).toBe("1");
    expect(change.length).toBe(5);
  }
  {
    const change = doc1.getChangeAtLamport("1", 10)!;
    expect(change.lamport).toBe(10);
    expect(change.peer).toBe("1");
    expect(change.length).toBe(5);
  }
  {
    const change = doc1.getChangeAtLamport("1", 13)!;
    expect(change.lamport).toBe(10);
    expect(change.peer).toBe("1");
    expect(change.length).toBe(5);
  }
  {
    const change = doc1.getChangeAtLamport("1", 20)!;
    expect(change.lamport).toBe(10);
    expect(change.peer).toBe("1");
    expect(change.length).toBe(5);
  }
  {
    const change = doc1.getChangeAtLamport("111", 13);
    expect(change).toBeUndefined();
  }
});

it("isContainer", () => {
  expect(isContainer("123")).toBeFalsy();
  expect(isContainer(123)).toBeFalsy();
  expect(isContainer(123n)).toBeFalsy();
  expect(isContainer(new Map())).toBeFalsy();
  expect(isContainer(new Set())).toBeFalsy();
  expect(isContainer({})).toBeFalsy();
  expect(isContainer(undefined)).toBeFalsy();
  expect(isContainer(null)).toBeFalsy();
  const doc = new LoroDoc();
  const t = doc.getText("t");
  expect(isContainer(t)).toBeTruthy();
  expect(isContainer(doc.getMap("m"))).toBeTruthy();
  expect(isContainer(doc.getMovableList("l"))).toBeTruthy();
  expect(isContainer(doc.getText("text"))).toBeTruthy();
  expect(isContainer(doc.getTree("tree"))).toBeTruthy();
  expect(isContainer(doc.getList("list"))).toBeTruthy();
  expect(getType(t)).toBe("Text");
  expect(getType(123)).toBe("Json");
});

it("getValueType", () => {
  // Type tests
  const doc = new LoroDoc();
  const t = doc.getText("t");
  expectTypeOf(getType(t)).toEqualTypeOf<"Text">();
  expect(getType(t)).toBe("Text");
  expectTypeOf(getType(123)).toEqualTypeOf<"Json">();
  expect(getType(123)).toBe("Json");
  expectTypeOf(getType(undefined)).toEqualTypeOf<"Json">();
  expect(getType(undefined)).toBe("Json");
  expectTypeOf(getType(null)).toEqualTypeOf<"Json">();
  expect(getType(null)).toBe("Json");
  expectTypeOf(getType({})).toEqualTypeOf<"Json">();
  expect(getType({})).toBe("Json");

  const map = doc.getMap("map");
  const list = doc.getList("list");
  const tree = doc.getTree("tree");
  const text = doc.getText("text");
  expectTypeOf(getType(map)).toEqualTypeOf<"Map">();
  expect(getType(map)).toBe("Map");
  expectTypeOf(getType(list)).toEqualTypeOf<"List">();
  expect(getType(list)).toBe("List");
  expectTypeOf(getType(tree)).toEqualTypeOf<"Tree">();
  expect(getType(tree)).toBe("Tree");
  expectTypeOf(getType(text)).toEqualTypeOf<"Text">();
  expect(getType(text)).toBe("Text");
});

it("enable timestamp", () => {
  const doc = new LoroDoc();
  doc.setPeerId(1);
  doc.getText("123").insert(0, "123");
  doc.commit();
  {
    const c = doc.getChangeAt({ peer: "1", counter: 0 });
    expect(c.timestamp).toBe(0);
  }

  doc.setRecordTimestamp(true);
  doc.getText("123").insert(0, "123");
  doc.commit();
  {
    const c = doc.getChangeAt({ peer: "1", counter: 4 });
    expect(c.timestamp).toBeCloseTo(Date.now() / 1000, -1);
  }
});

it("commit with specified timestamp", () => {
  const doc = new LoroDoc();
  doc.setPeerId(1);
  doc.getText("123").insert(0, "123");
  doc.commit({ timestamp: 111 });
  const c = doc.getChangeAt({ peer: "1", counter: 0 });
  expect(c.timestamp).toBe(111);
});

it("can control the mergeable interval", () => {
  {
    const doc = new LoroDoc();
    doc.setPeerId(1);
    doc.getText("123").insert(0, "1");
    doc.commit({ timestamp: 110 });
    doc.getText("123").insert(0, "1");
    doc.commit({ timestamp: 120 });
    expect(doc.getAllChanges().get("1")?.length).toBe(1);
  }

  {
    const doc = new LoroDoc();
    doc.setPeerId(1);
    doc.setChangeMergeInterval(10);
    doc.getText("123").insert(0, "1");
    doc.commit({ timestamp: 110 });
    doc.getText("123").insert(0, "1");
    doc.commit({ timestamp: 120 });
    console.log(doc.getAllChanges());
    expect(doc.getAllChanges().get("1")?.length).toBe(2);
  }
});

it("get container parent", () => {
  const doc = new LoroDoc();
  const m = doc.getMap("m");
  expect(m.parent()).toBeUndefined();
  const list = m.setContainer("t", new LoroList());
  expect(list.parent()!.id).toBe(m.id);
  const text = list.insertContainer(0, new LoroText());
  expect(text.parent()!.id).toBe(list.id);
  const tree = list.insertContainer(1, new LoroTree());
  expect(tree.parent()!.id).toBe(list.id);
  const treeNode = tree.createNode();
  const subtext = treeNode.data.setContainer("t", new LoroText());
  expect(subtext.parent()!.id).toBe(treeNode.data.id);
});

it("prelim support", () => {
  // Now we can create a new container directly
  const map = new LoroMap();
  map.set("3", 2);
  const list = new LoroList();
  list.insertContainer(0, map);
  // map should still be valid
  map.set("9", 9);
  // the type of setContainer/insertContainer changed
  const text = map.setContainer("text", new LoroText());
  {
    // Changes will be reflected in the container tree
    text.insert(0, "Heello");
    expect(list.toJSON()).toStrictEqual([{ "3": 2, "9": 9, text: "Heello" }]);
    text.delete(1, 1);
    expect(list.toJSON()).toStrictEqual([{ "3": 2, "9": 9, text: "Hello" }]);
  }
  const doc = new LoroDoc();
  const rootMap = doc.getMap("map");
  rootMap.setContainer("test", map); // new way to create sub-container

  // Use getAttached() to get the attached version of text
  const attachedText = text.getAttached()!;
  expect(text.isAttached()).toBeFalsy();
  expect(attachedText.isAttached()).toBeTruthy();
  text.insert(0, "Detached ");
  attachedText.insert(0, "Attached ");
  expect(text.toString()).toBe("Detached Hello");
  expect(doc.toJSON()).toStrictEqual({
    map: {
      test: {
        "3": 2,
        "9": 9,
        text: "Attached Hello",
      },
    },
  });
});

it("get elem by path", () => {
  const doc = new LoroDoc();
  const map = doc.getMap("map");
  map.set("key", 1);
  expect(doc.getByPath("map/key")).toBe(1);
  const map1 = doc.getByPath("map") as LoroMap;
  expect(getType(map1)).toBe("Map");
  map1.set("key1", 1);
  expect(doc.getByPath("map/key1")).toBe(1);
});

it("fork", () => {
  const doc = new LoroDoc();
  const map = doc.getMap("map");
  map.set("key", 1);
  const doc2 = doc.fork();
  const map2 = doc2.getMap("map");
  expect(map2.get("key")).toBe(1);
  expect(doc2.toJSON()).toStrictEqual({ map: { key: 1 } });
  map2.set("key", 2);
  expect(doc.toJSON()).toStrictEqual({ map: { key: 1 } });
  expect(doc2.toJSON()).toStrictEqual({ map: { key: 2 } });
  doc.import(doc2.exportSnapshot());
  expect(doc.toJSON()).toStrictEqual({ map: { key: 2 } });
});

describe("export", () => {
  it("test export update", () => {
    const doc = new LoroDoc();
    doc.getText("text").insert(0, "123");
    doc.commit();
    const updates = doc.export({ mode: "update", from: new VersionVector(null) });
    const doc2 = new LoroDoc();
    doc2.import(updates);
    expect(doc2.toJSON()).toStrictEqual({ text: "123" });
  })

  it("test export snapshot", () => {
    const doc = new LoroDoc();
    doc.getText("text").insert(0, "123");
    doc.commit();
    const snapshot = doc.export({ mode: "snapshot" });
    const doc2 = new LoroDoc();
    doc2.import(snapshot);
    expect(doc2.toJSON()).toStrictEqual({ text: "123" });
  })

  it("test export shallow-snapshot", () => {
    const doc = new LoroDoc();
    doc.getText("text").insert(0, "123");
    doc.commit();
    const snapshot = doc.export({ mode: "shallow-snapshot", frontiers: doc.oplogFrontiers() });
    const doc2 = new LoroDoc();
    doc2.import(snapshot);
    expect(doc2.toJSON()).toStrictEqual({ text: "123" });
  })

  it("test export updates-in-range", () => {
    const doc = new LoroDoc();
    doc.setPeerId(1);
    doc.getText("text").insert(0, "123");
    doc.commit();
    const bytes = doc.export({ mode: "updates-in-range", spans: [{ id: { peer: "1", counter: 0 }, len: 1 }] });
    const doc2 = new LoroDoc();
    doc2.import(bytes);
    expect(doc2.toJSON()).toStrictEqual({ text: "1" });
  })
})
it("has correct map value #453", async () => {
  {
    const doc = new LoroDoc();
    const text = doc.getText("text");
    text.insert(0, "Hello");
    text.mark({ start: 0, end: 2 }, "bold", { b: {} });
    expect(text.toDelta()).toStrictEqual([
      { insert: "He", attributes: { bold: { b: {} } } },
      { insert: "llo" }
    ]);
    let diff: Diff | undefined;
    let expectedDiff: TextDiff = {
      "type": "text",
      "diff": [
        { insert: "He", attributes: { bold: { b: {} } } },
        { insert: "llo" }
      ]
    };
    doc.subscribe(e => {
      console.log("Text", e);
      diff = e.events[0].diff;
    })
    doc.commit();
    await new Promise(resolve => setTimeout(resolve, 0));
    expect(diff).toStrictEqual(expectedDiff);
  }
  {
    const map = new LoroMap();
    map.set('a', { b: {} });
    expect(map.toJSON()).toStrictEqual({ a: { b: {} } });
  }
  {
    const doc = new LoroDoc();
    const map = doc.getMap("map");
    map.set('a', { b: {} });
    doc.commit();
    expect(map.toJSON()).toStrictEqual({ a: { b: {} } });
  }
  {
    const doc = new LoroDoc();
    let diff: Diff | undefined;
    const expectedDiff: MapDiff = {
      "type": "map",
      "updated": {
        "a": {
          "b": {}
        }
      }
    };
    doc.subscribe(e => {
      diff = e.events[0].diff;
    })
    const map = doc.getMap("map");
    map.set('a', { b: {} });
    doc.commit();
    await new Promise(resolve => setTimeout(resolve, 0));
    expect(diff).toStrictEqual(expectedDiff);
  }
})

it("can set commit message", () => {
  const doc = new LoroDoc();
  doc.setPeerId(1);
  doc.getText("text").insert(0, "123");
  doc.commit({ message: "Hello world" });
  expect(doc.getChangeAt({ peer: "1", counter: 0 }).message).toBe("Hello world");
})

it("can query pending txn length", () => {
  const doc = new LoroDoc();
  doc.setPeerId(1);
  expect(doc.getPendingTxnLength()).toBe(0);
  doc.getText("text").insert(0, "123");
  expect(doc.getPendingTxnLength()).toBe(3);
  doc.commit();
  expect(doc.getPendingTxnLength()).toBe(0);
})


it("can encode/decode frontiers", () => {
  const frontiers = [{ peer: "1123", counter: 1 }, { peer: "222", counter: 2 }] as Frontiers;
  const encoded = encodeFrontiers(frontiers);
  const decoded = decodeFrontiers(encoded);
  expect(decoded).toStrictEqual(frontiers);
})

it("travel changes", () => {
  let doc = new LoroDoc();
  doc.setPeerId(1);
  doc.getText("text").insert(0, "abc");
  doc.commit();
  let n = 0;
  doc.travelChangeAncestors([{ peer: "1", counter: 0 }], (meta: any) => {
    n += 1;
    return true
  })
  expect(n).toBe(1);
})

it("get path to container", () => {
  const doc = new LoroDoc();
  const map = doc.getMap("map");
  const list = map.setContainer("list", new LoroList());
  const path = doc.getPathToContainer(list.id);
  expect(path).toStrictEqual(["map", "list"])
})

it("json path", () => {
  const doc = new LoroDoc();
  const map = doc.getMap("map");
  map.set("key", "value");
  const books = map.setContainer("books", new LoroList());
  const book = books.insertContainer(0, new LoroMap());
  book.set("title", "1984");
  book.set("author", "George Orwell");
  const path = "$['map'].books[0].title";
  const result = doc.JSONPath(path);
  expect(result.length).toBe(1);
  expect(result).toStrictEqual(["1984"])
})

it("can push string to text", () => {
  const doc = new LoroDoc();
  const text = doc.getText("text");
  text.push("123");
  expect(text.toString()).toBe("123");
})

it("can push container to list", () => {
  const doc = new LoroDoc();
  const list = doc.getList("list");
  const map = list.pushContainer(new LoroMap());
  expect(list.toJSON()).toStrictEqual([{}]);
})

it("can push container to movable list", () => {
  const doc = new LoroDoc();
  const list = doc.getMovableList("list");
  const map = list.pushContainer(new LoroMap());
  expect(list.toJSON()).toStrictEqual([{}]);
})

it("can query the history for changed containers", () => {
  const doc = new LoroDoc();
  doc.setPeerId("0");
  doc.getText("text").insert(0, "H");
  doc.getMap("map").set("key", "H");
  const changed = doc.getChangedContainersIn({ peer: "0", counter: 0 }, 2)
  const changedSet = new Set(changed);
  expect(changedSet).toEqual(new Set([
    "cid:root-text:Text" as ContainerID,
    "cid:root-map:Map" as ContainerID,
  ]))
})

it("update VV", () => {
  const vv = new VersionVector(null);
  vv.setEnd({ peer: "1", counter: 1 });
  vv.setLast({ peer: "2", counter: 1 });
  vv.setLast({ peer: "3", counter: 4 });
  vv.remove("3");
  const map = vv.toJSON();
  expect(map).toStrictEqual(new Map([["1", 1], ["2", 2]]))
})
