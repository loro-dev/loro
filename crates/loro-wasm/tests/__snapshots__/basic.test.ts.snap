// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`can diff two versions 1`] = `
{
  "cid:root-list:List": {
    "diff": [
      {
        "insert": [
          "item1",
        ],
      },
    ],
    "type": "list",
  },
  "cid:root-map:Map": {
    "type": "map",
    "updated": {
      "key1": "value1",
      "key2": 42,
    },
  },
  "cid:root-text:Text": {
    "diff": [
      {
        "attributes": {
          "bold": true,
        },
        "insert": "Hello",
      },
    ],
    "type": "text",
  },
  "cid:root-tree:Tree": {
    "diff": [
      {
        "action": "create",
        "fractionalIndex": "80",
        "index": 0,
        "parent": undefined,
        "target": "12@1",
      },
      {
        "action": "create",
        "fractionalIndex": "80",
        "index": 0,
        "parent": "12@1",
        "target": "13@1",
      },
    ],
    "type": "tree",
  },
}
`;

exports[`can diff two versions 2`] = `
{
  "list": [
    "item1",
  ],
  "map": {
    "key1": "value1",
    "key2": 42,
  },
  "text": "Hello",
  "tree": [
    {
      "children": [
        {
          "children": [],
          "fractional_index": "80",
          "id": "11@2",
          "index": 0,
          "meta": {},
          "parent": "10@2",
        },
      ],
      "fractional_index": "80",
      "id": "10@2",
      "index": 0,
      "meta": {},
      "parent": null,
    },
  ],
}
`;

exports[`the diff will deduplication 1`] = `
{
  "cid:root-hi:Text": {
    "diff": [
      {
        "insert": "Hello",
      },
    ],
    "type": "text",
  },
  "cid:root-map:Map": {
    "type": "map",
    "updated": {
      "0": null,
      "1": null,
      "10": null,
      "11": null,
      "12": null,
      "13": null,
      "14": null,
      "15": null,
      "16": null,
      "17": null,
      "18": null,
      "19": null,
      "2": null,
      "20": null,
      "21": null,
      "22": null,
      "23": null,
      "24": null,
      "25": null,
      "26": null,
      "27": null,
      "28": null,
      "29": null,
      "3": null,
      "30": null,
      "31": null,
      "32": null,
      "33": null,
      "34": null,
      "35": null,
      "36": null,
      "37": null,
      "38": null,
      "39": null,
      "4": null,
      "40": null,
      "41": null,
      "42": null,
      "43": null,
      "44": null,
      "45": null,
      "46": null,
      "47": null,
      "48": null,
      "49": null,
      "5": null,
      "50": null,
      "51": null,
      "52": null,
      "53": null,
      "54": null,
      "55": null,
      "56": null,
      "57": null,
      "58": null,
      "59": null,
      "6": null,
      "60": null,
      "61": null,
      "62": null,
      "63": null,
      "64": null,
      "65": null,
      "66": null,
      "67": null,
      "68": null,
      "69": null,
      "7": null,
      "70": null,
      "71": null,
      "72": null,
      "73": null,
      "74": null,
      "75": null,
      "76": null,
      "77": null,
      "78": null,
      "79": null,
      "8": null,
      "80": null,
      "81": null,
      "82": null,
      "83": null,
      "84": null,
      "85": null,
      "86": null,
      "87": null,
      "88": null,
      "89": null,
      "9": null,
      "90": null,
      "91": null,
      "92": null,
      "93": null,
      "94": null,
      "95": null,
      "96": null,
      "97": null,
      "98": null,
      "99": null,
    },
  },
}
`;
